File: ports/go/src/core/thread/bcsthread.go
package thread

import (
	"sync"
	"time"
)

// BcsThread abstracts thread creation and joining, mapped to Goroutines
type BcsThread struct {
	wg       sync.WaitGroup
	quitChan chan struct{}
}

func NewBcsThread() *BcsThread {
	return &BcsThread{
		quitChan: make(chan struct{}),
	}
}

// Start executes the provided function in a new goroutine
func (t *BcsThread) Start(fn func()) {
	t.wg.Add(1)
	go func() {
		defer t.wg.Done()
		fn()
	}()
}

// Wait blocks until the spawned routine finishes
func (t *BcsThread) Wait() {
	t.wg.Wait()
}

// Sleep halts execution of the current routine for ms milliseconds
func Sleep(ms int) {
	time.Sleep(time.Duration(ms) * time.Millisecond)
}

File: ports/rust/src/core/thread/bcsthread.rs
use std::thread;
use std::time::Duration;

// BcsThread abstracts thread creation and joining for Rust
pub struct BcsThread {
    handle: Option<thread::JoinHandle<()>>,
}

impl BcsThread {
    pub fn new() -> Self {
        BcsThread { handle: None }
    }

    pub fn start<F>(&mut self, f: F)
    where
        F: FnOnce() + Send + 'static,
    {
        self.handle = Some(thread::spawn(f));
    }

    pub fn wait(&mut self) {
        if let Some(handle) = self.handle.take() {
            let _ = handle.join();
        }
    }
}

// sleep halts execution of the current thread for ms milliseconds
pub fn sleep(ms: u64) {
    thread::sleep(Duration::from_millis(ms));
}
